src/main/java/com/epam/aidial/core/controller/UploadFileController.java

```diff
@@ -29,25 +33,40 @@ protected Future<?> handle(ResourceDescription resource, boolean hasWriteAccess)
             return context.respond(HttpStatus.BAD_REQUEST, "Resource name and/or parent folders must not end with .(dot)");
         }
 
+        EtagHeader etag = EtagHeader.fromRequest(context.getRequest());
         Promise<Void> result = Promise.promise();
         context.getRequest()
                 .setExpectMultipart(true)
                 .uploadHandler(upload -> {
                     String contentType = upload.contentType();
                     Pipe<Buffer> pipe = new PipeImpl<>(upload).endOnFailure(false);
+                    FileService fileService = proxy.getFileService();
                     BlobWriteStream writeStream = new BlobWriteStream(
                             proxy.getVertx(),
                             proxy.getStorage(),
+                            fileService,
                             resource,
+                            etag,
                             contentType);
-                    pipe.to(writeStream, result);
+                    try {
+                        etag.validate(() -> fileService.getEtag(resource));
+                        pipe.to(writeStream, result);
+                    } catch (HttpException e) {
+                        result.fail(e);
+                    }
 
                     result.future()
-                            .onSuccess(success -> context.respond(HttpStatus.OK, writeStream.getMetadata()))
+                            .onSuccess(success -> {
+                                context.getResponse().putHeader(HttpHeaders.ETAG, writeStream.getMetadata().getEtag());
+                                context.respond(HttpStatus.OK, writeStream.getMetadata());
+                            })
                             .onFailure(error -> {
                                 writeStream.abortUpload(error);
-                                context.respond(HttpStatus.INTERNAL_SERVER_ERROR,
-                                        "Failed to upload file by path %s/%s".formatted(resource.getBucketName(), resource.getOriginalPath()));
+                                if (error instanceof HttpException exception) {
+                                    context.respond(exception.getStatus(), exception.getMessage());
```

Comment:
Seems like it could be refactored to ProxyContext#respond(Throwable error, String fallbackMessage)

Reply:
Done.